### PR TITLE
POL-749 Azure Hybrid Use Benefit for Windows/Linux Server Revamp

### PR DESCRIPTION
**Pull request by mburmeig**
### Description

This is a revamp of the Azure Hybrid Use Benefit for Windows Server policy as part of the general audit on usage and rate reduction policies. In addition to the usual updates to match functionality to the other revamped policies, the policy also now reports savings and has metadata to facilitate scraping into the optimization dashboard.

(Changes unrelated to reporting savings were also ported to the Azure Hybrid Use Benefit for Linux Server since the effort to do so was minimal. Additional discovery will be needed to determine if and how we can calculate AHUB savings for RHEL or SLES VMs.)

From the CHANGELOG:

- Several parameters altered to be more descriptive and human-readable
- Added ability to only report recommendations that meet a minimum savings threshold
- Added ability to use Subscription list parameter as either an "allow" list or a "deny" list
- Added ability to filter resources by region
- Added ability to filter resources by multiple tag key:value pairs
- Added additional context to incident description
- Normalized incident export to be consistent with other policies
- Added human-readable recommendation to incident export
- Added estimated savings to incident export
- Added additional fields to incident export to facilitate scraping for dashboards
- Streamlined code for better readability and faster execution

### Link to Example Applied Policy

https://app.flexera.com/orgs/6/automation/applied-policies/projects/7954?policyId=65049d696a972a0001e50559

### Contribution Check List

- [X] New functionality includes testing.
- [X] New functionality has been documented in the README if applicable
- [X] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/azure/hybrid_use_benefit/README.md [hunk 1]
```diff
@@ -2,13 +2,14 @@
...
 
 ## Functional Details
 
-- The policy identifies all Windows server instances that are not currently using [Azure Hybrid Use Benefit](https://azure.microsoft.com/en-us/pricing/hybrid-benefit/). It raises an incident for all applicable VMs not currently using AHUB, which once approved, will enable AHUB on all identified instances.1
+- The policy identifies all Windows server instances that are not currently using [Azure Hybrid Use Benefit](https://azure.microsoft.com/en-us/pricing/hybrid-benefit/). It raises an incident for all applicable VMs not currently using AHUB and provides a monthly savings amount if AHUB is enabled, which once approved, will enable AHUB on all identified instances.
 - The Exclusion Tag parameter is a string value. Supply the Tag Key only. Tag Values are not analyzed and therefore are not need. If the exclusion tag key is used on an Instance, that Instance is presumed to be exempt from this policy.
 - This policy does not track licenses or availability. It is your responsibility to ensure you are not under licensed.
+- The savings currency is provided in $USD, and currency conversion is not supported.
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/README.md:12]: We should add support for this. We now have an API capable of delivering this. See this policy: https://raw.githubusercontent.com/flexera-public/policy_templates/POL-832-aws-unused-ip-audit/cost/aws/unused_ip_addresses/aws_unused_ip_addresses.pt

Specifically ds_currency_target and the following datasources. You should be able to basically copypaste this into this policy and tweak it to match what this policy is doing.

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 2]
```diff
@@ -7,9 +7,9 @@ severity "low"
 category "Cost"
 default_frequency "daily"
 info(
-  version: "2.12",
+  version: "3.0",
   provider: "Azure",
-  service: "",
+  service: "compute",
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:12]: compute should be capitalized to match other policies.

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 3]
```diff
@@ -132,14 +132,23 @@ datasource "azure_instances" do
     end
 end
 
+datasource "datasource_ds_azure_pricing" do
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:135]: Datasource names should begin with ds_, not datasource_ds_. See:

https://flexera.atlassian.net/wiki/spaces/FP/pages/2184642622/Policy+Best+Practices#Block-Naming-Conventions

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 4]
```diff
@@ -160,40 +169,102 @@ EOS
...
 script "js_filter_resources", type: "javascript" do
-  parameters "azure_instances","exclusion_tag"
+  parameters "azure_instances","exclusion_tag", "pricing"
   result "result"
   code <<-EOS
     var result = [];
-    _.each(azure_instances, function(instance){
-      if (_.has(instance.tags, exclusion_tag)) {
-        console.log(instance["name"], "SKIP")
+    var region_table = {
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:176]: Add spaces between keys and values, like so: "ukwest": "UK West"

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 5]
```diff
@@ -160,40 +169,102 @@ EOS
...
-          }
+        return null;
+      }
+    }
+
+    _.each(azure_instances, function(value) {
+      var regionMapping = region_table[value.location];
+      var priceDifference = getPriceInfo(regionMapping, value.vmSize, value.osType, pricing);
+        if (value.osType == "Windows" && priceDifference != null && value.licenseType == null){
+          monthlySavings = ((priceDifference * 24) * 30 )
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:250]: To match parity with some changes we are making to our savings policies, the math here should be slightly different. Replace 30 with 365.25 / 12.

Reasoning: 365.25 is the average number of days in a year (due to a leap year every 4 years), so this gives us a much closer approximation of what the average monthly cost would be.

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 6]
```diff
@@ -132,14 +132,23 @@ datasource "azure_instances" do
...
+    header "User-Agent", "RS Policies"
+  end
+end
+
 datasource "filtered_resources" do
-  run_script $js_filter_resources, $azure_instances, $param_exclusion_tag_key
+  run_script $js_filter_resources, $azure_instances, $param_exclusion_tag_key, $datasource_ds_azure_pricing
 end
 
 ##########
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:147]: Datasources and Scripts should now be in a single combined section. I realize you didn't make it this way, but this is one of those things where, cuz you're touching the policy, it's yours now to fix :)

https://flexera.atlassian.net/wiki/spaces/FP/pages/2184642622/Policy+Best+Practices#Basic-Policy-Structure

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 7]
```diff
@@ -160,40 +169,102 @@ EOS
 end
 
 script "js_filter_resources", type: "javascript" do
-  parameters "azure_instances","exclusion_tag"
+  parameters "azure_instances","exclusion_tag", "pricing"
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:172]: parameter variable names should be identical to their datasource or parameter counterparts when applicable. See: https://flexera.atlassian.net/wiki/spaces/FP/pages/2184642622/Policy+Best+Practices#Scripts

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 8]
```diff
@@ -223,6 +294,15 @@ policy 'azure_license_policy' do
       field "name" do
```
Comment (XOmniverse) [inline cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:294]: Incident output should have both subscription name and subscription id, and they should match accountname and accountid like our other policies to ensure the results are properly scraped. See the incident output section of this policy and try to include everything that would also make sense in this policy, with the same names:

https://raw.githubusercontent.com/flexera-public/policy_templates/master/cost/azure/old_snapshots/azure_delete_old_snapshots.pt

**Review by XOmniverse — CHANGES_REQUESTED** [review]

**Review by XOmniverse — CHANGES_REQUESTED** [review]
Here's my next set of recommendations. Let me know once these are in place.

File: cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt [hunk 9]
```diff
@@ -79,6 +86,7 @@ end
 # Datasources
```
[the remark this answers is not in the document]
Comment (mburmeig) [reply cost/azure/hybrid_use_benefit/azure_hybrid_use_benefit.pt:86]: Made recommended changes.

**Review by XOmniverse — APPROVED** [review]

**Review by taterdog590 — APPROVED** [review]
LGTM

**Pull request merged**